BaseRecalibrator, optimized

jean-philippe-martin opened the pull request.

davidadamsphd commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/DataflowCommandLineProgram.java:
```diff
@@ -85,19 +128,24 @@ public static String getRunnerTypeName(String runnerSimpleName) {
+            && (projectID == null || stagingLocation==null)) {
             throw new UserException.CommandLineException(String.format("Non local dataflow execution requires project " +
-                    "and staging to be set. project:%s id:%s.",projectID, stagingLocation));
+                    "and staging to be set. project:%s stagine:%s.",projectID, stagingLocation));
```
staging location

jean-philippe-martin commented:
I'm pushing an updated version in a minute that's rebased, in the hope that will fix the sequencedictionary-related test failures.

droazen commented on src/main/java/org/broadinstitute/hellbender/bench/LoadRecalTablea.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ *
+ */
+public class LoadRecalTablea {
```
Make this into a proper hellbender `CommandLineProgram` with the program group in its `CommandLineProgramProperties` annotation set to a new `BenchmarkingProgramGroup`

jean-philippe-martin replied:
Removed.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/DataflowCommandLineProgram.java:
```diff
@@ -131,19 +138,24 @@ public static String getRunnerTypeName(String runnerSimpleName) {
 
+    @Argument(doc = "The Google Compute Engine machine type that Dataflow will use when spinning up worker VMs",
+         fullName = "workerMachineType", optional=true)
+    protected String workerMachineType = "n1-standard-1";
```
Is this defaulting to a 1 core machine only because of the BQSR bug?

jean-philippe-martin replied:
yes, exactly.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/DataflowCommandLineProgram.java:
```diff
@@ -131,19 +138,24 @@ public static String getRunnerTypeName(String runnerSimpleName) {
         return null;
     }
 
+    public GenomicsFactory.OfflineAuth offlineAuth;
```
Move field declaration to top of class and make field private (providing a public getter if necessary)

jean-philippe-martin replied:
OK. I removed it instead, as now we can get the authentication info from the pipeline options. This is a lot more convenient.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/AddContextDataToShardedReads.java:
```diff
@@ -0,0 +1,35 @@
+
+/**
+ *
+ */
```
Add class-level documentation

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/AddContextDataToShardedReads.java:
```diff
@@ -0,0 +1,35 @@
+        SimpleInterval shard = new SimpleInterval(shardedReads.getKey());
+        // how many extra bases to grab on either side.
+        // make sure we're not going past the end of the contig.
+        int margin = 100;
```
Margin should be a static constant declared at top of class

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/AddContextDataToShardedReads.java:
```diff
@@ -0,0 +1,35 @@
+        int margin = 100;
+        // 1-based
+        int start = shard.getStart()-margin;
+        if (start<1) start=1;
```
`Math.max(shard.getStart() - margin, 1)`

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/AddContextDataToShardedReads.java:
```diff
@@ -0,0 +1,35 @@
+        final int refContigLength = refDictionary.getSequence(shard.getContig()).getSequenceLength();
+        // included
+        int end = shard.getEnd()+margin;
+        if (end>refContigLength) end = refContigLength;
```
`Math.min(shard.getEnd() + margin, refContigLength)`

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/AddContextDataToShardedReads.java:
```diff
@@ -0,0 +1,35 @@
+        if (end>refContigLength) end = refContigLength;
+
+
+        SimpleInterval paddedShard = new SimpleInterval(shard.getContig(), start, end);
```
Extract static utility method `padInterval(interval, padding, contigLength)` to `IntervalUtils`, and call it here

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/AddContextDataToShardedReads.java:
```diff
@@ -0,0 +1,35 @@
+        ReferenceBases referenceBases = refSource.getReferenceBases(opts, shard);
+
+        return new ReadContextData(referenceBases, null);
+    }
```
Realized after reviewing this class that it's not used anywhere -- delete!

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReadsDataflowSource.java:
```diff
@@ -142,6 +143,22 @@ public SAMFileHeader getHeader() {
+    /**
+      * Create a {@link PCollection< KV <String,Iterable<GATKRead>>>} containing all the reads starting in the given intervals and their mates,
+      * grouped by the shard they start in.
+      * The shards are strings as per SimpleInterval's ToString syntax.
```
Don't use the `toString()` representation as the key (`toString()` is just for display purposes) -- use the `SimpleIntervals` themselves.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReadsDataflowSource.java:
```diff
@@ -142,6 +143,22 @@ public SAMFileHeader getHeader() {
+      * Create a {@link PCollection< KV <String,Iterable<GATKRead>>>} containing all the reads starting in the given intervals and their mates,
+      * grouped by the shard they start in.
+      * The shards are strings as per SimpleInterval's ToString syntax.
+      * Reads that are malformed or unmapped are ignored.
```
Define malformed in this context.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/RefAPISource.java:
```diff
@@ -30,7 +31,7 @@
 import java.security.GeneralSecurityException;
 import java.util.*;
-
+import java.util.concurrent.TimeUnit;
```
Remove unused imports

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+/**
+ * see ShardedReadsDatasource.get
+ */
+public final class ShardedReadsDatasource {
```
Name this `ShardedReadsDataflowSource` for consistency with the surrounding classes.

droazen replied:
Turns out this entire class may be unused -- if it is, please delete.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                    //    + " cutShards: " + cutShards.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " kv.of: " + kving.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " output: " + outputting.elapsed(TimeUnit.MILLISECONDS) + ", "
+                    //    + " rejected " + rejected);
```
Delete commented-out code

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                                if (!ret.isEmpty()) {
+                                    accepted += ret.size();
+                                    choosing.stop();
+                                    shipping.start();
```
Remove stopwatch code from the version going into master, as it makes the class illegible

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+        Collections.shuffle(shards);
+        logger.info("Input range sharded into "+shards.size()+" intervals.");
+        List<KV<String,Void>> intervalKeys;
+        intervalKeys = shards.stream().map(s-> KV.<String,Void>of(s.toString(), null)).collect(Collectors.<KV<String,Void>>toList());
```
Use intervals directly as the keys instead of `toString()` representation

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                        // local file, use setIntervalsForTraversal
+                        querying.start();
+                        ReadsDataSource rds = new ReadsDataSource(new File(bam));
+                        rds.setIntervalsForTraversal(is);
```
`Use rds.query(interval)` instead, since there's only one interval here.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                    readingInput.stop();
+                    Stopwatch init = Stopwatch.createStarted();
+                    ArrayList<SimpleInterval> is = new ArrayList<>();
+                    is.add(i);
```
Choose a more descriptive name than `is` for this List

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                    if (!cloudStorageUrl && !hadoopUrl) {
+                        // local file, use setIntervalsForTraversal
+                        querying.start();
+                        ReadsDataSource rds = new ReadsDataSource(new File(bam));
```
You are not closing this data source -- wrap in try-with-resources

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                    List<GATKRead> ret = new ArrayList<GATKRead>();
+                    init.stop();
+
+                    if (!cloudStorageUrl && !hadoopUrl) {
```
When loading from a local file, it looks like you fail to set the header in the underlying SAMRecords to null. If the method claims to strip the header from the reads, it should do so in all cases.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                    } else if (cloudStorageUrl) {
+                        // GCS file, load just the bit we need.
+                        querying.start();
+                        final SamReader reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.DEFAULT_STRINGENCY);
```
You are not closing this reader -- try-with-resources.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+     * The shards are strings as per SimpleInterval's ToString syntax.
+     * Reads that are malformed or unmapped are ignored.
+     *
+     * Note: for performance reasons, this will give you SAMRecords that have no header. You have to put it back in.
```
This warning needs to be way more prominent -- the reads will be returned in a potentially unusable state for some operations until the header is restored (at least until we patch htsjdk as planned in https://github.com/broadinstitute/hellbender/issues/900 to make headerless operation an officially supported mode).

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                                    continue;
+                                }
+
+                                while (currentShard.getEnd() < r.getStart()) {
```
There's a lot of duplicated code between the local and cloud storage cases -- can you extract a common method?

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                        loading.stop();
+                    } else {
+                        // TODO: HDFS input
+                        throw new RuntimeException("Sorry, Hadoop support not yet implemented");
```
Document this limitation in the class comments

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+                    }
+                }
+
+                private boolean accept(SAMRecord r, SimpleInterval region) {
```
Can you get rid of this overload, and get both cases calling into the `GATKRead` version (perhaps by reordering where the adapter is applied to the read?)

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+
+
+    // chr2:1-200 -> chr2:1-100,chr2:101-200
+    static public List<SimpleInterval> cutToShards(Iterable<SimpleInterval> intervals, int shardSize) {
```
Add docs for public methods (here and below).

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+            ret.add(new SimpleInterval(i.getContig(), beginOfShard(endShard, shardSize), i.getEnd()));
+        }
+        return ret;
+    }
```
This `cutToShards()` methods strikes me as exactly the kind of method that would have edge-case bugs -- could you add good unit tests for this method in isolation?

jean-philippe-martin replied:
done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+     * Each key corresponds to a shard (1-based shards)
+     * Each Locatable is placed in each shard that it overlaps (that means they may be placed more than once).
+     */
+    static public HashMap<String, ArrayList<Locatable>> groupByOverlappingShard(Iterable<? extends Locatable> intervals, int shardSize) {
```
Should this (and the methods below) really be public?

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+            for (int index = startIndex; index<=endIndex; index++) {
+                SimpleInterval startInterval = new SimpleInterval(i.getContig(), beginOfShard(index, shardSize), endOfShard(index, shardSize));
+                String key = startInterval.toString();
+                if (!ret.containsKey(key)) ret.put(key, new ArrayList<>());
```
Add braces around if body as security against future bugs.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+     * Iterator of T -> iterator of arrays of T (by reading multiple at once).
+     * The last array may be smaller, as we run out of elements.
+     */
+    private static final class BatchingIterator<T> implements Iterator<ArrayList<T>> {
```
Extract to a public class in `utils.iterators`, and add a basic unit test

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+    private static final class BatchingIterator<T> implements Iterator<ArrayList<T>> {
+        final Iterator<T> iter;
+        final int count;
+        ArrayList<T> buf;
```
Make fields private

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+        }
+
+        // if you pass it an array, it'll be used to put new content. Don't access it until
+        // you get it back via next().
```
Make these comments into actual javadoc comments

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java:
```diff
@@ -0,0 +1,382 @@
+            this.iter = iter;
+            this.count = count;
+            this.buf = buf;
+            if (buf==null) this.buf = new ArrayList<T>(count);
```
Add braces around all if/else statements (here and below).

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+         * creates a new shard, adding the specified variants.
+         */
+        public ContextShard withVariants(ArrayList<Variant> newVariants) {
+            return new ContextShard(this.interval, new IntervalsSkipList<>(newVariants), reads, readContext);
```
Will the object be in a consistent state after this call? Won't the variants in `readContext` be out of sync with `newVariants`?

jean-philippe-martin replied:
ContextShard is just a holder class, it enforces no consistency guarantees. I've added a note to this effect.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                if (BucketUtils.isCloudStorageUrl(bam)) {
+                    storageClient = DataflowCommandLineProgram.HellbenderDataflowOptions.Methods.createStorageClient(c.getPipelineOptions().as(DataflowCommandLineProgram.HellbenderDataflowOptions.class));
+                    reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.SILENT);
+                } else {
```
What about the `BucketUtils.isHadoopUrl()` case? Do you want to at least throw in that case if it's not supported (and document in the class docs that it's unsupported)?

jean-philippe-martin replied:
done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                ContextShard shard = c.element();
+                ArrayList<SimpleInterval> ints =new ArrayList<>();
+                ints.add(shard.interval);
+                List<SimpleInterval> subshards = ShardedReadsDatasource.cutToShards(ints, outputShardSize);
```
Aha, so `cutToShards()` at least is still used -- in that case, my earlier comment about standalone unit tests for that method still applies.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+     *
+     * @param bam URL to the reads
+     * @param outputShardSize subsharding size
+     * @param optFilter if specified, only reads that satisfy this will be included.
```
Missing doc for the margin arg

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                            }
+                            if (r.getAlignmentStart()<firstValidPos) {
+                                throw new GATKException("Margin was too tight, a read starts early by "+(shard.interval.getStart()-r.getAlignmentStart())+", starting at "+r.getAlignmentStart());
+                            }
```
Why are you only checking the read against `lastValidPos` and `firstValidPos` if it's unmapped? This makes no sense...shouldn't you be checking *mapped* reads against the shard boundaries?

jean-philippe-martin replied:
Ah of course, that was backwards. Thanks!

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                    while (query.hasNext()) {
+                        SAMRecord r = query.next();
+                        if (!accept(r, shard.interval)) continue;
+                        if (r.getReadUnmappedFlag()) {
```
I think you should be wrapping in the read adapter earlier here so that we can use the definition of unmapped provided by the `GATKRead` interface (which covers more cases than just having the unmapped flag set).

jean-philippe-martin replied:
ok!

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                                readsSoFar = new ArrayList<>();
+                            }
+                            // move to the next shard
+                            currentSubShard = subshards.get(++currentSubShardIndex);
```
Check that `currentSubShardIndex` does not go out of bounds as part of your loop condition?

jean-philippe-martin replied:
that should never happen, since the shard only contains read that start inside it. If this went out of bounds it would indicate a bug in the code, and so a stack trace is the right thing to send.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+    /**
+     * Fill in reads that start in the given shard, and subshard the output to the requested size.
+     * Signals an error if any read sticks out more than "margin" outside of the big shard.
+     * Optionally filters the reads to only keep the ones that satisfy the given predicate.
```
Document the fact that the reads are headerless.

jean-philippe-martin replied:
I made it nicely prominent, I think you'll like it.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                super.finishBundle(c);
+            }
+
+            private boolean accept(SAMRecord r, SimpleInterval region) {
```
Recommend that you merge `accept()` with the test against `optFilter`, and make accept() operate on a `GATKRead` by wrapping in the adapter a bit earlier (this will simplify the method quite a bit, since `GATKRead.isUnmapped()` does all of the unmapped checks below for you).

jean-philippe-martin replied:
I know. I did that for performance (skipping the allocation if this isn't a read we want to keep), but will undo it for you.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                getRef = Stopwatch.createUnstarted();
+                queryVariants = Stopwatch.createUnstarted();
+                sending = Stopwatch.createUnstarted();
+                total = Stopwatch.createUnstarted();
```
Remove `Stopwatch` code from the master branch version to improve readability (unless you have objections).

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                elementCount++;
+                ContextShard shard = c.element();
+                SAMSequenceDictionary refDict = c.sideInput(refDictView);
+                int margin = 500;
```
Declare margin as static constant at top of class

droazen replied:
Actually, shouldn't you be using the `BQSR_REFERENCE_WINDOW_FUNCTION` to determine the reference bases for each read instead of this fixed constant? 500 is not always theoretically enough if there are extremely long clipped sequences around a read.

jean-philippe-martin replied:
We could, but using a constant is a lot faster than calling a function on every read. Is there a constant we could use that you would be comfortable with?

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                int margin = 500;
+                int start = shard.interval.getStart();
+                if (start>margin) start -= margin;
+                else start = 1; // it's 1-based, remember.
```
Replace with `int start = Math.max(shard.interval.getStart() - margin, 1);`

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                int len = refDict.getSequence(shard.interval.getContig()).getSequenceLength();
+                getRef.stop();
+                if (end<len-margin) end+= margin;
+                else end = len; // OK because those are 1-based closed intervals
```
Replace the above with `int end = Math.min(shard.interval.getEnd() + margin, refDict.getSequence(shard.interval.getContig()).getSequenceLength());`

jean-philippe-martin replied:
Well, the existing code protects us from overflows in case end is close to maxint. You're right, though, we can do without and min is easier to read.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+
+                ArrayList<ReadContextData> readContext = new ArrayList<>();
+                for (GATKRead r : shard.reads) {
+                    SimpleInterval readInterval = new SimpleInterval(r.getContig(), r.getStart(), r.getEnd());
```
Simplify to `new SimpleInterval(r)`

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+                    queryVariants.stop();
+                    // we pass all the bases. That's better because this way it's just a shared
+                    // pointer instead of being an array copy. Downstream processing is fine with having
+                    // extra bases (it expects a few, actually).
```
Again, see my note about `BQSR_REFERENCE_WINDOW_FUNCTION` above -- you might in theory not have all the bases you need here...

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow.java:
```diff
@@ -175,8 +177,13 @@ public void processElement(ProcessContext c) throws Exception {
 
+            // save post-skel reads, for debugging
+            //PCollection<GATKRead> readsWithoutContext = readsWithContext.apply(Keys.create());
+            //SmallBamWriter.writeToFile(pipeline, readsWithoutContext, readsHeader, "tmp_post-skel-reads.bam");
```
Remove commented-out code

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow.java:
```diff
@@ -186,7 +193,7 @@ public void processElement(ProcessContext c) {
                     }
                 }));
-
+            */
```
Is the commented-out block here no longer needed? If not, remove

jean-philippe-martin replied:
yup, removed.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+        oneLineSummary = "Generates recalibration table",
+        programGroup = ReadProgramGroup.class
+)
+public class BaseRecalibratorDataflow2 extends DataflowCommandLineProgram implements Serializable {
```
Rename to `BaseRecalibratorDataflowOptimized`

jean-philippe-martin replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,330 @@
+ *
+ * In the end we have reads, variants, and reference bases.
+ */
+public class AddContextDataToReadOptimized {
```
`AddContextDataToReadOptimized` should have standalone unit tests like the non-optimized version.

jean-philippe-martin replied:
That will take a little while. Please remind me in case I forget by the time I push the next revision.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+        try {
+            bunny.start("BaseRecalibratorDataflow2");
+
+            if (null==baseRecalibrationKnownVariants || baseRecalibrationKnownVariants.size()==0) {
```
`|| baseRecalibrationKnownVariants.isEmpty()`

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+            bunny.stepEnd("set up bam input");
+
+            // Load the Variants and the Reference
+            //final VariantsDataflowSource variantsDataflowSource = new VariantsDataflowSource(baseRecalibrationKnownVariants, pipeline);
```
Remove commented-out line

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+            // the granularity at which we'll want to assign work and read inputs.
+            final int bigShardSize = 1_000_000;
+            // the granularity at which we want to batch reads,variants,and reference for processing.
+            final int smallShardSize = 5_000;
```
Declare these constants at the top of the class instead of embedded within the code.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+            List<Variant> variants = VariantsDataflowSource.getVariantsList(baseRecalibrationKnownVariants);
+
+            bunny.stepEnd("load variants");
+            int margin = 1000;
```
declare as constant at top of class, with explanatory comment

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+                .apply(ParDo.named("subdivideAndFillReads").of(AddContextDataToReadOptimized.subdivideAndFillReads(bam, smallShardSize, margin, filterToApply)))
+                // add ref bases to the shards
+                .apply(ParDo.named("fillContext").of(AddContextDataToReadOptimized.fillContext(referenceDataflowSource, refDictionaryView, offlineAuth))
+                    .withSideInputs(refDictionaryView));
```
Can you encapsulate the 3-4 smaller operations involved in adding context data to the reads (ie., `fillVariants()`, `subdivideAndFillReads()`, `fillContext()`, etc.) within a single top-level method `AddContextDataToReadOptimized.add()` that we can call to do everything?

jean-philippe-martin replied:
I can do it for you, sure. It'll be an awkwardly large function, though, even more so because it mixes client code and worker code. 
And it muddles the AddContextDataToReadOpt class which used to be purely worker code, to also contain client code.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+            //PCollection<KV<GATKRead, ReadContextData>> readsWithContext = shardsWithContext
+            //    // convert to the old format so the rest of the code can work.
+            //    .apply(ParDo.named("flattenShards").of(AddContextDataToRead2.flattenShards(BQSR_REFERENCE_WINDOW_FUNCTION)));
+            //final PCollection<RecalibrationTables> recalibrationTable = readsWithContext.apply(new BaseRecalibratorTransform(headerSingleton, refDictionaryView, BRAC));
```
Remove commented-out code

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+
+    // Given a list of shards and a list of variants,
+    // add each variant to every (shard+margin) that it overlaps
+    private ArrayList<AddContextDataToReadOptimized.ContextShard> fillVariants(List<SimpleInterval> shardedIntervals, List<Variant> variants, int margin) {
```
Move to the `AddContextDataToReadOptimized` class

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2.java:
```diff
@@ -0,0 +1,283 @@
+        IntervalsSkipList<Variant> fastVars = new IntervalsSkipList<>(variants);
+        for (SimpleInterval s : shardedIntervals) {
+            int start = s.getStart() - margin;
+            if (start<1) start=1;
```
`Math.max(s.getStart() - margin, 1)`

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorFn.java:
```diff
@@ -191,10 +200,16 @@ private void apply(GATKRead originalRead, final ReferenceDataSource refDS, Itera
+        //final int[] isSNP = calculateIsSNP(read, refDS);
+        //final int[] isInsertion = calculateIsIndel(read, EventType.BASE_INSERTION);
+        //final int[] isDeletion = calculateIsIndel(read, EventType.BASE_DELETION);
+        //final int nErrors = nEvents(isSNP, isInsertion, isDeletion);
```
The minor refactoring you did in this class conflicts with the BQSR engine refactoring we did in https://github.com/broadinstitute/hellbender/tree/dr_lb_spark_bqsr -- if there are any changes here that must be preserved for **performance** reasons, please let us know.

jean-philippe-martin replied:
grouping those functions together should give us a tiny little bit of performance. I would like this preserved, though I haven't measured the performance impact yet.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorOptimizedFn.java:
```diff
@@ -0,0 +1,572 @@
+ * DoFn for BaseRecalibrator.
+ * Takes in reads + contextual data (overlapping reference bases and variants), spits out RecalibrationTables.
+ */
+public final class BaseRecalibratorOptimizedFn extends DoFnWLog<AddContextDataToReadOptimized.ContextShard, RecalibrationTables> {
```
According to my diff tool, this class is almost identical to `BaseRecalibratorFn` above -- when we rebase https://github.com/broadinstitute/hellbender/tree/dr_lb_spark_bqsr on top of this, `BaseRecalibratorFn` and `BaseRecalibratorOptimizedFn` will both become thin shells that call into a shared `BaseRecalibrationEngine`. If there's any subtle optimization in this class that I'm not seeing, please let me know so that I can preserve it during the rebase.

jean-philippe-martin replied:
Correct, the only point here is to work on a shard instead of individual reads. Re-creating the BaseRecalibrationEngine is the right thing to do.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorOptimizedTransform.java:
```diff
@@ -0,0 +1,155 @@
+                }
+            }
+        }));
+        */
```
Remove commented-out block

jean-philippe-martin replied:
Are you sure? As per instructions, the code still isn't correct. We may need this code back very soon.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorTransform.java:
```diff
@@ -40,13 +46,36 @@ public BaseRecalibratorTransform(final PCollectionView<SAMFileHeader> headerView
+                }
+            }
+        }));
+        */
```
Remove commented-out block

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorTransform.java:
```diff
@@ -102,6 +131,15 @@ public void processElement(ProcessContext c) throws IOException {
+                        int numReadGroups = header.getReadGroups().size();
+                        RecalibrationEngine recalibrationEngine = new RecalibrationEngine(covariates, numReadGroups);
+                        rt = recalibrationEngine.getRecalibrationTables();
+                        RecalibrationEngine.finalizeRecalibrationTables(rt);
```
Is this zero reads case covered by tests?

jean-philippe-martin replied:
It isn't yet.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/recalibration/RecalDatum.java:
```diff
@@ -15,6 +15,7 @@
     public static final byte MAX_RECALIBRATED_Q_SCORE = SAMUtils.MAX_PHRED_SCORE;
     private static final double UNINITIALIZED = -1.0;
     private static final long serialVersionUID = 1L;
+    private static final double MULTIPLIER = 1000.0;
```
The changes here to `RecalDatum` are in https://github.com/broadinstitute/hellbender/pull/878, right? Are you going to merge that version first, and rebase this one away?

jean-philippe-martin replied:
Yes.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/bqsr/RecalibrationEngine.java:
```diff
@@ -66,8 +67,17 @@ public void updateDataForRead( final ReadRecalibrationInfo recalInfo ) {
+                    for (int j=0; j<2; j++) {
+                        if (keys[j]<0) {
+                            throw new GATKException.ShouldNeverReachHereException("keys["+j+"]="+keys[j]+", should not have been negative.");
+                        }
```
This is to help track down the BQSR threading bug, I imagine?

jean-philippe-martin replied:
yup. I can remove it from master if you want, that one isn't hard to recreate :)

droazen replied:
Can be removed now that we've identified the culprit

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java:
```diff
@@ -209,7 +209,24 @@ public final boolean overlaps( final Locatable other ) {
+             return false;
+         }
+
+         return this.contig.equals(other.getContig()) && this.start <= other.getEnd() + margin && other.getStart() <= this.end + margin;
```
Can you write the last part as `other.getStart() - margin <= this.end`? It seems a little more logical that way.

droazen replied:
Also add a basic unit test for this method.

jean-philippe-martin replied:
I prefer it the other way, but I'll change it for you. I've added the test, too.

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/dataflow/datasources/RefAPISourceUnitTest.java:
```diff
@@ -37,12 +42,24 @@ private ReferenceBases queryReferenceAPI( final String referenceName, final Simp
+        if (getDataflowTestApiKey() != null) {
+            options.setApiKey(getDataflowTestApiKey());
+        }
+        if (getDataflowTestApiKey()!=null) {
```
Combine the two consecutive `if (getDataflowTestApiKey()!=null)` statements.

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflow2IntegrationTest.java:
```diff
@@ -0,0 +1,259 @@
+ *                          - human_g1k_v37.chr17_1Mb.*
+ *                          - NA12878.chr17_69k_70k.dictFix.*
+ */
+public final class BaseRecalibratorDataflow2IntegrationTest extends CommandLineProgramTest {
```
Rename to `BaseRecalibratorDataflowOptimizedIntegrationTest`

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflowIntegrationTest.java:
```diff
@@ -111,11 +111,13 @@ public void runAllTests() throws Exception {
         final String hg19Ref = "EMWV_ZfLxrDY-wE";
         final String GRCh37Ref = RefAPISource.GRCH37_REF_ID;
         final String HiSeqBam = localResources + "CEUTrio.HiSeq.WGS.b37.ch20.1m-1m1k.NA12878.bam";
+        final String trickyBam = localResources + "CEUTrio.HiSeq.WGS.b37.ch20.4379150-4379157.bam";
```
Why is this a `trickyBam`? Can you add an explanatory comment?

jean-philippe-martin replied:
It's called tricky because that's the one that made even BaseRecalibrator give wrong answers.
See https://github.com/broadinstitute/hellbender/pull/878 for the full background.

droazen commented:
Review complete! Looks very good overall, but as usual had a bunch of comments. Back to @jean-philippe-martin

droazen commented:
Do we know why tests are failing on this branch, by the way?

jean-philippe-martin commented:
I'm looking into it. Looks like a combination of missing files and at least one actual bug that creeped in after the "perf first" command.

jean-philippe-martin commented:
@droazen I have added the tests you were asking for and changed the reference base code to the slower but safer version (plus this version should hopefully pass the tests now). Let me know if this looks good to you! Then I can rebase and merge.

droazen commented on src/main/java/local/LoadRecalTable.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * temporary debug thing
+ */
+public class LoadRecalTable {
```
This was supposed to have been removed?

jean-philippe-martin replied:
src/main/java/local was supposed to be in .gitignore... wait, it is (?)

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,380 @@
+     * Takes the variants, groups them into shards of size "bigShardSize" at the client, then
+     * runs subdivideAndFillReads then fillContext. See their documentation for more information.
+     */
+    public static PCollection<ContextShard> subdivideAndFillEverything(
```
Can you name this top-level entry point something less awkward like AddContextDataToReadOptimized.add(), and supply an overload with fewer parameters that uses sensible defaults for bigShardSize, outputShardSize, margin, and marginForRef? We want this common operation of "adding overlapping data to a read" to be simple to use, if possible.

jean-philippe-martin replied:
Sure!

droazen replied:
(and actually call the top-level method from BQSR, to ensure that it doesn't rot)

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java:
```diff
@@ -0,0 +1,380 @@
+     * Given a shard that has reads and variants, query Google Genomics' Reference server and get reference info
+     * (including an extra margin on either side), and fill that and the correct variants into readContext.
+     */
+    public static DoFn<ContextShard, ContextShard> fillContext(final ReferenceDataflowSource refSource, final PCollectionView<SAMSequenceDictionary> refDictView, int margin) throws IOException {
```
Isn't the `margin` argument ignored now that you're computing the exact reference bases required for the shard?

jean-philippe-martin replied:
Ah yes, good point. I'll remove it here and at the callers.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/dataflow/ShardUtils.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * chr2:1-200 -> chr2:1-100,chr2:101-200
+     */
+    static public List<SimpleInterval> cutToShards(Iterable<SimpleInterval> intervals, int shardSize) {
```
This should really be in `IntervalUtils`, as it's not dataflow-specific.

jean-philippe-martin replied:
Done!

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/collections/IntervalsSkipListOneContigTest.java:
```diff
@@ -0,0 +1,130 @@
+/**
+ *
+ */
+public class IntervalsSkipListOneContigTest {
```
Should extend `BaseTest` and name should end in `UnitTest`

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/collections/IntervalsSkipListTest.java:
```diff
@@ -9,68 +11,34 @@
 /**
  *
  */
 public class IntervalsSkipListTest {
```
Should extend `BaseTest` and name should end in `UnitTest`

droazen commented:
Second-pass review complete -- feel free to rebase and merge after addressing remaining few comments, and once tests pass.

jean-philippe-martin commented:
It appears the merge process has undone a bunch of the work. Working on fixing that now :-(

droazen commented:
Ouch! Recommend that you do a force-reset back to the previous state to be certain nothing is lost.
 `git reset --hard 03e77c14c36c8b7b5ab4182d72260c2f8e41780b`

jean-philippe-martin commented:
OK, but then what? Rebasing again will almost certainly return to the same outcome.

droazen commented:
After the reset, do a local rebase and squash first. Ie., 
`git reset --hard 03e77c14c36c8b7b5ab4182d72260c2f8e41780b`
then
`git rebase -i 5fc5ca9f3a5bbe76b147b82ade0b7e00968d812f^`
(and squash into one commit during this step)

Then force push, and attempt the rebase onto origin/master again.

droazen commented:
Note that the `^` at the end of the second command is important!

jean-philippe-martin commented:
gotta love git.

jean-philippe-martin commented:
Tests are green. Merging, with fingers crossed.

akiezun commented:
@jean-philippe-martin please delete merged branches
